In [16]:
import numpy as np
import pandas as pd
from sklearn.model_selection import cross_val_score

In [17]:
df=pd.read_csv('D:\Data Science\Big_Mart_Sales_Prediction\Preprocessed_Data.csv')
df.head()

<>:1: SyntaxWarning: invalid escape sequence '\D'
<>:1: SyntaxWarning: invalid escape sequence '\D'
C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15192\3931347399.py:1: SyntaxWarning: invalid escape sequence '\D'
  df=pd.read_csv('D:\Data Science\Big_Mart_Sales_Prediction\Preprocessed_Data.csv')


,Unnamed: 0,Item_Weight,Item_Visibility,Item_MRP,Outlet_Establishment_Year,Item_Outlet_Sales,Outlet_Establishment_Age,Item_Fat_Content_Low Fat,Item_Fat_Content_Regular,Outlet_Type_Grocery Store,...,Item_Type_Health and Hygiene,Item_Type_Household,Item_Type_Meat,Item_Type_Others,Item_Type_Seafood,Item_Type_Snack Foods,Item_Type_Soft Drinks,Item_Type_Starchy Foods,Outlet Size,Outlet Location Type
0,0,9.30,0.252232,249.8092,1999,2.796007,25,1.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0
1,1,5.92,0.268136,48.2692,2009,2.142162,15,0.0,1.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,2.0
2,2,17.50,0.255913,141.6180,1999,2.601398,25,1.0,0.0,0.0,...,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0
3,3,19.20,0.436908,182.0950,1998,2.280825,26,0.0,1.0,1.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0
4,4,8.93,0.393053,53.8614,1987,2.369801,37,1.0,0.0,0.0,...,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,2.0


In [18]:
df.columns

Index(['Unnamed: 0', 'Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year', 'Item_Outlet_Sales',
       'Outlet_Establishment_Age', 'Item_Fat_Content_Low Fat',
       'Item_Fat_Content_Regular', 'Outlet_Type_Grocery Store',
       'Outlet_Type_Supermarket Type1', 'Outlet_Type_Supermarket Type2',
       'Outlet_Type_Supermarket Type3', 'Item_Type_Baking Goods',
       'Item_Type_Breads', 'Item_Type_Breakfast', 'Item_Type_Canned',
       'Item_Type_Dairy', 'Item_Type_Frozen Foods',
       'Item_Type_Fruits and Vegetables', 'Item_Type_Hard Drinks',
       'Item_Type_Health and Hygiene', 'Item_Type_Household', 'Item_Type_Meat',
       'Item_Type_Others', 'Item_Type_Seafood', 'Item_Type_Snack Foods',
       'Item_Type_Soft Drinks', 'Item_Type_Starchy Foods', 'Outlet Size',
       'Outlet Location Type'],
      dtype='object')

In [19]:
x=df.drop(['Item_Outlet_Sales','Unnamed: 0','Outlet_Establishment_Year','Item_Fat_Content_Low Fat','Item_Type_Baking Goods','Outlet_Type_Grocery Store'],axis='columns')
y=df['Item_Outlet_Sales']

In [20]:
x

,Item_Weight,Item_Visibility,Item_MRP,Outlet_Establishment_Age,Item_Fat_Content_Regular,Outlet_Type_Supermarket Type1,Outlet_Type_Supermarket Type2,Outlet_Type_Supermarket Type3,Item_Type_Breads,Item_Type_Breakfast,...,Item_Type_Health and Hygiene,Item_Type_Household,Item_Type_Meat,Item_Type_Others,Item_Type_Seafood,Item_Type_Snack Foods,Item_Type_Soft Drinks,Item_Type_Starchy Foods,Outlet Size,Outlet Location Type
0,9.300,0.252232,249.8092,25,0.0,1.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0
1,5.920,0.268136,48.2692,15,1.0,0.0,1.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,2.0
2,17.500,0.255913,141.6180,25,0.0,1.0,0.0,0.0,0.0,0.0,...,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0
3,19.200,0.436908,182.0950,26,1.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0
4,8.930,0.393053,53.8614,37,0.0,1.0,0.0,0.0,0.0,0.0,...,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,2.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
8374,6.865,0.384362,214.5218,37,0.0,1.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,2.0,2.0
8375,8.380,0.360838,108.1570,22,1.0,1.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0
8376,10.600,0.327686,85.1224,20,0.0,1.0,0.0,0.0,0.0,0.0,...,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0
8377,7.210,0.525625,103.1332,15,1.0,0.0,1.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,2.0


In [21]:
x.columns

Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Age', 'Item_Fat_Content_Regular',
       'Outlet_Type_Supermarket Type1', 'Outlet_Type_Supermarket Type2',
       'Outlet_Type_Supermarket Type3', 'Item_Type_Breads',
       'Item_Type_Breakfast', 'Item_Type_Canned', 'Item_Type_Dairy',
       'Item_Type_Frozen Foods', 'Item_Type_Fruits and Vegetables',
       'Item_Type_Hard Drinks', 'Item_Type_Health and Hygiene',
       'Item_Type_Household', 'Item_Type_Meat', 'Item_Type_Others',
       'Item_Type_Seafood', 'Item_Type_Snack Foods', 'Item_Type_Soft Drinks',
       'Item_Type_Starchy Foods', 'Outlet Size', 'Outlet Location Type'],
      dtype='object')

In [22]:
y

0       2.796007
1       2.142162
2       2.601398
3       2.280825
4       2.369801
          ...   
8374    2.694474
8375    2.200264
8376    2.424294
8377    2.560160
8378    2.293534
Name: Item_Outlet_Sales, Length: 8379, dtype: float64

### **Model Fitting**

#### **1. Linear Regression**

In [23]:
from sklearn.linear_model import LinearRegression
lr=LinearRegression()
score_lr=cross_val_score(lr,X=x,y=y,cv=5)

In [24]:
print(f"Score for Linear Regression Model : {np.mean(score_lr)}")

Score for Linear Regression Model : 0.6971927712665351


#### **2. Lasso Regression**

In [25]:
from sklearn.linear_model import Lasso
from sklearn.model_selection import GridSearchCV
lasso_parameter={
    'alpha':[0.0001,0.0005,0.0009,0.01,0.05,0.09,0.1,0.5,0.9]
}
lasso_reg=GridSearchCV(Lasso(),param_grid=lasso_parameter,cv=5,n_jobs=-1,verbose=1,return_train_score=False)
lasso_reg.fit(x,y)

Fitting 5 folds for each of 9 candidates, totalling 45 fits


GridSearchCV(cv=5, estimator=Lasso(), n_jobs=-1,
             param_grid={'alpha': [0.0001, 0.0005, 0.0009, 0.01, 0.05, 0.09,
                                   0.1, 0.5, 0.9]},
             verbose=1)

In [26]:
print(f'Best score : {lasso_reg.best_score_}')
print(f'Best Parameter : {lasso_reg.best_params_}')

Best score : 0.697390032524218
Best Parameter : {'alpha': 0.0001}


#### **3. Ridge Regression**

In [27]:
from sklearn.linear_model import Ridge
ridge_parameters={
    'alpha':[0.0001,0.0005,0.0009,0.01,0.05,0.09,0.1,0.5,0.9]
}
ridge_reg=GridSearchCV(Ridge(),param_grid=ridge_parameters,cv=5,n_jobs=-1,verbose=1,return_train_score=False)
ridge_reg.fit(x,y)

Fitting 5 folds for each of 9 candidates, totalling 45 fits


GridSearchCV(cv=5, estimator=Ridge(), n_jobs=-1,
             param_grid={'alpha': [0.0001, 0.0005, 0.0009, 0.01, 0.05, 0.09,
                                   0.1, 0.5, 0.9]},
             verbose=1)

In [28]:
print(f'Best Score : {ridge_reg.best_score_}')
print(f'Best Parameter : {ridge_reg.best_params_}')

Best Score : 0.6971939154062579
Best Parameter : {'alpha': 0.5}


#### **4. Random Forest Regressor**

In [29]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import RandomizedSearchCV

rf_parameter={
    'n_estimators':[100,150,200,250],
    'max_depth': [5,6,7,8],
    'min_samples_split': [8,9,10,11],
    'min_samples_leaf': [20,22,24,25]
}
random_forest=RandomizedSearchCV(RandomForestRegressor(),param_distributions=rf_parameter,n_iter=100,verbose=2,return_train_score=False)
random_forest.fit(x,y)

Fitting 5 folds for each of 100 candidates, totalling 500 fits
[CV] END max_depth=5, min_samples_leaf=20, min_samples_split=8, n_estimators=250; total time=   5.7s
[CV] END max_depth=5, min_samples_leaf=20, min_samples_split=8, n_estimators=250; total time=   5.1s
[CV] END max_depth=5, min_samples_leaf=20, min_samples_split=8, n_estimators=250; total time=   5.1s
[CV] END max_depth=5, min_samples_leaf=20, min_samples_split=8, n_estimators=250; total time=   5.1s
[CV] END max_depth=5, min_samples_leaf=20, min_samples_split=8, n_estimators=250; total time=   5.1s
[CV] END max_depth=7, min_samples_leaf=22, min_samples_split=8, n_estimators=250; total time=   6.7s
[CV] END max_depth=7, min_samples_leaf=22, min_samples_split=8, n_estimators=250; total time=   6.5s
[CV] END max_depth=7, min_samples_leaf=22, min_samples_split=8, n_estimators=250; total time=   6.6s
[CV] END max_depth=7, min_samples_leaf=22, min_samples_split=8, n_estimators=250; total time=   6.6s
[CV] END max_depth=7, min_sa

RandomizedSearchCV(estimator=RandomForestRegressor(), n_iter=100,
                   param_distributions={'max_depth': [5, 6, 7, 8],
                                        'min_samples_leaf': [20, 22, 24, 25],
                                        'min_samples_split': [8, 9, 10, 11],
                                        'n_estimators': [100, 150, 200, 250]},
                   verbose=2)

In [30]:
print(f'Best Score : {random_forest.best_score_}')
print(f'Best Parameter : {random_forest.best_params_}')

Best Score : 0.713341987908632
Best Parameter : {'n_estimators': 100, 'min_samples_split': 9, 'min_samples_leaf': 25, 'max_depth': 6}


#### **XGBoost Regressor**

In [36]:
from xgboost import XGBRegressor
from sklearn.model_selection import RandomizedSearchCV
xgb_params={
    'n_estimators':[12,13,15,16,17,18,19,20,22,24,25],
    'max_depth':[4,5,6,7],
    'booster':['gbtree'],
    'min_child_weight':[60,65,72,75,76,78,80],
    'base_score':[3,4,5,6,7,8],
    'lambda':[2,4,6,8,10,12],
    'alpha':[3,6,8,9,12,14]
}
xgb=RandomizedSearchCV(XGBRegressor(),param_distributions=xgb_params,cv=8,n_iter=100,n_jobs=-1,verbose=2,return_train_score=False)
xgb.fit(X=x,y=y)

Fitting 8 folds for each of 100 candidates, totalling 800 fits


RandomizedSearchCV(cv=8,
                   estimator=XGBRegressor(base_score=None, booster=None,
                                          callbacks=None,
                                          colsample_bylevel=None,
                                          colsample_bynode=None,
                                          colsample_bytree=None, device=None,
                                          early_stopping_rounds=None,
                                          enable_categorical=False,
                                          eval_metric=None, feature_types=None,
                                          feature_weights=None, gamma=None,
                                          grow_policy=None,
                                          importance_type=None,
                                          interaction_constraint...
                                          monotone_constraints=None,
                                          multi_strategy=None,
                                          n_estimators=None, n_jobs=None,
                                          num_parallel_tree=None, ...),
                   n_iter=100, n_jobs=-1,
                   param_distributions={'alpha': [3, 6, 8, 9, 12, 14],
                                        'base_score': [3, 4, 5, 6, 7, 8],
                                        'booster': ['gbtree'],
                                        'lambda': [2, 4, 6, 8, 10, 12],
                                        'max_depth': [4, 5, 6, 7],
                                        'min_child_weight': [60, 65, 72, 75, 76,
                                                             78, 80],
                                        'n_estimators': [12, 13, 15, 16, 17, 18,
                                                         19, 20, 22, 24, 25]},
                   verbose=2)

In [37]:
print(f'Best Score : {xgb.best_score_}')
print(f'Best Parameters : {xgb.best_params_}')

Best Score : 0.712149655795211
Best Parameters : {'n_estimators': 25, 'min_child_weight': 75, 'max_depth': 6, 'lambda': 2, 'booster': 'gbtree', 'base_score': 7, 'alpha': 3}


From the above model We can clearly see that Random Forest Regressor method gives the highest R2 score 71.33% among all the models.
We will select Random Forest Regressor {n_estimators': 100, 'min_samples_split': 9, 'min_samples_leaf': 25, 'max_depth': 6} for this dataset

In [40]:
import pickle
file=open('randomforest_model.pkl','wb')
pickle.dump(random_forest,file)